gcp/rakefiles/secrets.rb

```diff
@@ -3,37 +3,74 @@
 
 class Secrets
 
-  def self.set_secrets()
-    saved_secrets_file_path = "../#{ENV['ENV']}/secrets/#{ENV["TF_VAR_project_id"]}-secrets.yml"
-
-    begin
-      secrets = YAML.load(File.read(saved_secrets_file_path))
-    rescue Errno::ENOENT
-      generate_file = true
-      secrets = Hash.new
-    end
+  KMS_KEYRING = "keyring"
+  KMS_DEFAULT_KEY = "default"
 
-    [ \
-      'couchdb_admin_username', \
-      'couchdb_admin_password', \
-      'couchdb_secret', \
-    ].each do |secret|
-      unless ENV[secret.upcase].to_s.empty?
-        secrets[secret] = ENV[secret.upcase]
-        # we don't want to store Environment variables
-        generate_file = false
-      end
-      secrets[secret] = SecureRandom.hex if secrets[secret].to_s.empty?
-      ENV["TF_VAR_#{secret}"] = secrets[secret]
-    end
+  SECRETS_FILE = "secrets.yaml"
+
+  def self.collect_secrets()
+    ENV['TF_VAR_keyring_name'] = Secrets::KMS_KEYRING
+    $compose_env << "TF_VAR_keyring_name"
 
-    if generate_file
-      puts "Secret file #{saved_secrets_file_path} for this deployment not found. I will create one."
-      File.open(saved_secrets_file_path, 'w+') do |file|
-        file.write(secrets.to_yaml)
+    encryption_keys = []
+    secrets = {}
+    %x{find ../../modules -maxdepth 2 | grep #{Secrets::SECRETS_FILE}}.split.each do |mod|
+      mod_default_key = File.dirname(mod).split('/').last
+      mod_secrets_cfg = YAML.load(File.read(mod))
+
+      kms_key = mod_secrets_cfg['kms_key'] ? mod_secrets_cfg['kms_key'] : mod_default_key
+      encryption_keys << "\"#{kms_key}\"" unless encryption_keys.include? kms_key
+
+      if secrets[kms_key]
+        secrets[kms_key].concat(mod_secrets_cfg['secrets'])
+      else
+        secrets[kms_key] = mod_secrets_cfg['secrets']
+      end
+      mod_secrets_cfg['secrets'].each do |secret|
+        $compose_env << "TF_VAR_#{secret}"
       end
     end
 
+    ENV["TF_VAR_encryption_keys"] = %Q|[ #{encryption_keys.join(", ")} ]|
+    $compose_env << "TF_VAR_encryption_keys"
+
     return secrets
   end
+
+  def self.set_secrets(secrets)
+    if secrets
+      secrets.each do |key, secrets|
+        sh_filter "#{$exekube_cmd} secrets-fetch #{key}"
+
+        secrets_file = "./secrets/#{key}/#{Secrets::SECRETS_FILE}"
+        if File.file?(secrets_file)
+          secrets = YAML.load(File.read(secrets_file))
```

Comment:
Oh no, a third definition of `secrets`!

How about `populated_secrets`, since this is the same data structure as the one in `populated_secrets` in the `else` clause of this stanza?

Reply:
Done